File: android/src/main/java/com/elsapiens/backgroundlocation/TrackingStateStore.java
package com.elsapiens.backgroundlocation;

/**
 * Persists the desired tracking state across process death.
 *
 * The services and restart receivers consult this store to decide whether a restart is
 * wanted and with which parameters. Without it, a system-initiated restart used to run
 * with a bogus "default_reference"/"auto_restart" reference, silently corrupting the
 * recorded route. Single responsibility: state persistence only — no Android service
 * logic lives here.
 */
public class TrackingStateStore {

    // Task tracking keys
    private static final String KEY_TASK_ACTIVE = "task_active";
    private static final String KEY_TASK_REFERENCE = "task_reference";
    private static final String KEY_TASK_INTERVAL = "task_interval";
    private static final String KEY_TASK_MIN_DISTANCE = "task_min_distance";
    private static final String KEY_TASK_HIGH_ACCURACY = "task_high_accuracy";
    private static final String KEY_TASK_MAX_ACCURACY = "task_max_accuracy";
    private static final String KEY_NOTIFICATION_TITLE = "notification_title";
    private static final String KEY_NOTIFICATION_TEXT = "notification_text";

    // Work hour tracking keys
    private static final String KEY_WORK_HOUR_ACTIVE = "work_hour_active";
    private static final String KEY_WORK_HOUR_ENGINEER_ID = "work_hour_engineer_id";
    private static final String KEY_WORK_HOUR_UPLOAD_INTERVAL = "work_hour_upload_interval";
    private static final String KEY_WORK_HOUR_SERVER_URL = "work_hour_server_url";
    private static final String KEY_WORK_HOUR_AUTH_TOKEN = "work_hour_auth_token";
    private static final String KEY_WORK_HOUR_OFFLINE_QUEUE = "work_hour_offline_queue";

    public static final long DEFAULT_INTERVAL_MS = 3000L;
    public static final float DEFAULT_MIN_DISTANCE_METERS = 10.0f;
    public static final float DEFAULT_MAX_ACCURACY_METERS = 30.0f;
    public static final long DEFAULT_UPLOAD_INTERVAL_MS = 300000L;

    private final KeyValueStore store;

    public TrackingStateStore(KeyValueStore store) {
        this.store = store;
    }

    // ------------------------------------------------------------------
    // Task tracking state
    // ------------------------------------------------------------------

    public void saveTaskTracking(TaskTrackingState state) {
        store.putBoolean(KEY_TASK_ACTIVE, true);
        store.putString(KEY_TASK_REFERENCE, state.reference);
        store.putLong(KEY_TASK_INTERVAL, state.interval);
        store.putFloat(KEY_TASK_MIN_DISTANCE, state.minDistance);
        store.putBoolean(KEY_TASK_HIGH_ACCURACY, state.highAccuracy);
        store.putFloat(KEY_TASK_MAX_ACCURACY, state.maxAccuracy);
        store.putString(KEY_NOTIFICATION_TITLE, state.notificationTitle);
        store.putString(KEY_NOTIFICATION_TEXT, state.notificationText);
    }

    public void clearTaskTracking() {
        store.putBoolean(KEY_TASK_ACTIVE, false);
        store.remove(KEY_TASK_REFERENCE);
    }

    public boolean isTaskTrackingActive() {
        return store.getBoolean(KEY_TASK_ACTIVE, false);
    }

    /** Returns the persisted task state, or null when tracking is not active. */
    public TaskTrackingState getTaskTracking() {
        if (!isTaskTrackingActive()) {
            return null;
        }
        String reference = store.getString(KEY_TASK_REFERENCE, null);
        if (reference == null || reference.trim().isEmpty()) {
            return null;
        }
        return new TaskTrackingState(
            reference,
            store.getLong(KEY_TASK_INTERVAL, DEFAULT_INTERVAL_MS),
            store.getFloat(KEY_TASK_MIN_DISTANCE, DEFAULT_MIN_DISTANCE_METERS),
            store.getBoolean(KEY_TASK_HIGH_ACCURACY, true),
            store.getFloat(KEY_TASK_MAX_ACCURACY, DEFAULT_MAX_ACCURACY_METERS),
            store.getString(KEY_NOTIFICATION_TITLE, null),
            store.getString(KEY_NOTIFICATION_TEXT, null)
        );
    }

    // ------------------------------------------------------------------
    // Work hour tracking state
    // ------------------------------------------------------------------

    public void saveWorkHourTracking(WorkHourState state) {
        store.putBoolean(KEY_WORK_HOUR_ACTIVE, true);
        store.putString(KEY_WORK_HOUR_ENGINEER_ID, state.engineerId);
        store.putLong(KEY_WORK_HOUR_UPLOAD_INTERVAL, state.uploadInterval);
        store.putString(KEY_WORK_HOUR_SERVER_URL, state.serverUrl);
        store.putString(KEY_WORK_HOUR_AUTH_TOKEN, state.authToken);
        store.putBoolean(KEY_WORK_HOUR_OFFLINE_QUEUE, state.enableOfflineQueue);
    }

    public void clearWorkHourTracking() {
        store.putBoolean(KEY_WORK_HOUR_ACTIVE, false);
        store.remove(KEY_WORK_HOUR_ENGINEER_ID);
        store.remove(KEY_WORK_HOUR_SERVER_URL);
        store.remove(KEY_WORK_HOUR_AUTH_TOKEN);
    }

    public boolean isWorkHourTrackingActive() {
        return store.getBoolean(KEY_WORK_HOUR_ACTIVE, false);
    }

    /** Returns the persisted work-hour state, or null when tracking is not active. */
    public WorkHourState getWorkHourTracking() {
        if (!isWorkHourTrackingActive()) {
            return null;
        }
        String engineerId = store.getString(KEY_WORK_HOUR_ENGINEER_ID, null);
        String serverUrl = store.getString(KEY_WORK_HOUR_SERVER_URL, null);
        if (engineerId == null || serverUrl == null) {
            return null;
        }
        return new WorkHourState(
            engineerId,
            store.getLong(KEY_WORK_HOUR_UPLOAD_INTERVAL, DEFAULT_UPLOAD_INTERVAL_MS),
            serverUrl,
            store.getString(KEY_WORK_HOUR_AUTH_TOKEN, null),
            store.getBoolean(KEY_WORK_HOUR_OFFLINE_QUEUE, true)
        );
    }

    // ------------------------------------------------------------------
    // Value objects
    // ------------------------------------------------------------------

    public static class TaskTrackingState {
        public final String reference;
        public final long interval;
        public final float minDistance;
        public final boolean highAccuracy;
        public final float maxAccuracy;
        public final String notificationTitle;
        public final String notificationText;

        public TaskTrackingState(String reference, long interval, float minDistance, boolean highAccuracy,
                float maxAccuracy, String notificationTitle, String notificationText) {
            this.reference = reference;
            this.interval = interval;
            this.minDistance = minDistance;
            this.highAccuracy = highAccuracy;
            this.maxAccuracy = maxAccuracy;
            this.notificationTitle = notificationTitle;
            this.notificationText = notificationText;
        }
    }

    public static class WorkHourState {
        public final String engineerId;
        public final long uploadInterval;
        public final String serverUrl;
        public final String authToken;
        public final boolean enableOfflineQueue;

        public WorkHourState(String engineerId, long uploadInterval, String serverUrl, String authToken,
                boolean enableOfflineQueue) {
            this.engineerId = engineerId;
            this.uploadInterval = uploadInterval;
            this.serverUrl = serverUrl;
            this.authToken = authToken;
            this.enableOfflineQueue = enableOfflineQueue;
        }
    }
}

File: ios/Sources/BackgroundLocationPlugin/TrackingStateStore.swift
import Foundation

/// Persists the desired tracking state across process death and relaunch.
///
/// iOS can terminate a suspended app at any time and later relaunch it in the
/// background (in response to a significant location change, for instance) with no
/// memory of prior state. This store is what lets `TaskLocationTracker` and
/// `WorkHourLocationTracker` resume the right session — with the right reference,
/// interval and accuracy — instead of starting a bogus one. Mirrors the Android
/// plugin's `TrackingStateStore` field-for-field. Single responsibility: persistence
/// only — no CoreLocation logic lives here.
final class TrackingStateStore {

    // Task tracking keys
    private static let keyTaskActive = "task_active"
    private static let keyTaskReference = "task_reference"
    private static let keyTaskInterval = "task_interval"
    private static let keyTaskMinDistance = "task_min_distance"
    private static let keyTaskHighAccuracy = "task_high_accuracy"
    private static let keyTaskMaxAccuracy = "task_max_accuracy"

    // Work hour tracking keys
    private static let keyWorkHourActive = "work_hour_active"
    private static let keyWorkHourEngineerId = "work_hour_engineer_id"
    private static let keyWorkHourUploadInterval = "work_hour_upload_interval"
    private static let keyWorkHourServerUrl = "work_hour_server_url"
    private static let keyWorkHourAuthToken = "work_hour_auth_token"
    private static let keyWorkHourOfflineQueue = "work_hour_offline_queue"

    static let defaultIntervalSeconds: Double = 3.0
    static let defaultMinDistanceMeters: Double = 10.0
    static let defaultMaxAccuracyMeters: Double = 30.0
    static let defaultUploadIntervalSeconds: Double = 300.0

    private let store: KeyValueStore

    init(store: KeyValueStore) {
        self.store = store
    }

    // MARK: Task tracking state

    struct TaskTrackingState {
        let reference: String
        let interval: Double
        let minDistance: Double
        let highAccuracy: Bool
        let maxAccuracy: Double
    }

    func saveTaskTracking(_ state: TaskTrackingState) {
        store.set(true, forKey: Self.keyTaskActive)
        store.set(state.reference, forKey: Self.keyTaskReference)
        store.set(state.interval, forKey: Self.keyTaskInterval)
        store.set(state.minDistance, forKey: Self.keyTaskMinDistance)
        store.set(state.highAccuracy, forKey: Self.keyTaskHighAccuracy)
        store.set(state.maxAccuracy, forKey: Self.keyTaskMaxAccuracy)
    }

    func clearTaskTracking() {
        store.set(false, forKey: Self.keyTaskActive)
        store.removeObject(forKey: Self.keyTaskReference)
    }

    var isTaskTrackingActive: Bool {
        store.bool(forKey: Self.keyTaskActive)
    }

    /// The persisted task session, or nil when tracking is not active or the
    /// reference is missing/blank (guards against resuming a bogus session).
    func getTaskTracking() -> TaskTrackingState? {
        guard isTaskTrackingActive,
              let reference = store.string(forKey: Self.keyTaskReference),
              !reference.trimmingCharacters(in: .whitespaces).isEmpty else {
            return nil
        }
        let interval = store.double(forKey: Self.keyTaskInterval)
        let minDistance = store.double(forKey: Self.keyTaskMinDistance)
        let maxAccuracy = store.double(forKey: Self.keyTaskMaxAccuracy)
        return TaskTrackingState(
            reference: reference,
            interval: interval > 0 ? interval : Self.defaultIntervalSeconds,
            minDistance: minDistance > 0 ? minDistance : Self.defaultMinDistanceMeters,
            highAccuracy: store.bool(forKey: Self.keyTaskHighAccuracy),
            maxAccuracy: maxAccuracy > 0 ? maxAccuracy : Self.defaultMaxAccuracyMeters
        )
    }

    // MARK: Work hour tracking state

    struct WorkHourState {
        let engineerId: String
        let uploadInterval: Double
        let serverUrl: String
        let authToken: String?
        let enableOfflineQueue: Bool
    }

    func saveWorkHourTracking(_ state: WorkHourState) {
        store.set(true, forKey: Self.keyWorkHourActive)
        store.set(state.engineerId, forKey: Self.keyWorkHourEngineerId)
        store.set(state.uploadInterval, forKey: Self.keyWorkHourUploadInterval)
        store.set(state.serverUrl, forKey: Self.keyWorkHourServerUrl)
        store.set(state.authToken, forKey: Self.keyWorkHourAuthToken)
        store.set(state.enableOfflineQueue, forKey: Self.keyWorkHourOfflineQueue)
    }

    func clearWorkHourTracking() {
        store.set(false, forKey: Self.keyWorkHourActive)
        store.removeObject(forKey: Self.keyWorkHourEngineerId)
        store.removeObject(forKey: Self.keyWorkHourServerUrl)
        store.removeObject(forKey: Self.keyWorkHourAuthToken)
    }

    var isWorkHourTrackingActive: Bool {
        store.bool(forKey: Self.keyWorkHourActive)
    }

    /// The persisted work-hour session, or nil when inactive or missing required fields.
    func getWorkHourTracking() -> WorkHourState? {
        guard isWorkHourTrackingActive,
              let engineerId = store.string(forKey: Self.keyWorkHourEngineerId),
              let serverUrl = store.string(forKey: Self.keyWorkHourServerUrl) else {
            return nil
        }
        let uploadInterval = store.double(forKey: Self.keyWorkHourUploadInterval)
        return WorkHourState(
            engineerId: engineerId,
            uploadInterval: uploadInterval > 0 ? uploadInterval : Self.defaultUploadIntervalSeconds,
            serverUrl: serverUrl,
            authToken: store.string(forKey: Self.keyWorkHourAuthToken),
            enableOfflineQueue: store.bool(forKey: Self.keyWorkHourOfflineQueue)
        )
    }
}
